### PR TITLE
Add category

### PR DESCRIPTION
**Pull request by AGuliaiev**

### CODE REVIEW

File: src/main/java/org/example/springintro/dto/book/CreateBookRequestDto.java [hunk 1]
```diff
@@ -23,4 +24,6 @@ public class CreateBookRequestDto {
     private String description;
     @NotBlank
     private String coverImage;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/dto/book/CreateBookRequestDto.java]:
```suggestion
    @NotEmpty
```

File: src/main/java/org/example/springintro/model/Book.java [hunk 2]
```diff
@@ -34,4 +42,13 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/org/example/springintro/model/Book.java [hunk 3]
```diff
@@ -34,4 +42,13 @@ public class Book {
...
+    @JoinTable(
+            name = "book_categories",
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/model/Book.java]:
```suggestion
            name = "books_categories",
```

File: src/main/java/org/example/springintro/model/Category.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+    @ManyToMany(mappedBy = "categories", cascade = CascadeType.REMOVE)
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    private Set<Book> books = new HashSet<>();
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/model/Category.java]:
```suggestion
```

File: src/main/java/org/example/springintro/services/impl/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override
+    public CategoryDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toEntity(categoryDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public CategoryDto updateById(Long id, CategoryDto categoryDto) {
+        Category category = findByIdOrThrow(id);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/services/impl/CategoryServiceImpl.java:41]:
```suggestion
        Category category = findByIdOrThrow(id);
        mapper.updateCategoryFromDto(categoryDto, category);
```

like example

File: src/main/java/org/example/springintro/services/impl/CategoryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
+    }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> findBooksByCategoryId(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springintro/services/impl/CategoryServiceImpl.java]: it should be in BookController and BookService

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by okuzan — APPROVED** [review]

**Pull request merged**